halo2/src/recursion/endoscale/primitive.rs

```diff
@@ -0,0 +1,161 @@
+//! Primitives used in endoscaling.
+
+use group::{Curve, Group};
+use pasta_curves::arithmetic::{CurveAffine, FieldExt};
+
+use subtle::CtOption;
+
+/// Maps a pair of bits to a multiple of a scalar using endoscaling.
+pub(crate) fn endoscale_pair_scalar<F: FieldExt>(bits: [bool; 2]) -> F {
+    // [2 * bits.0 - 1]
+    let mut scalar = F::from(bits[0]).double() - F::one();
+
+    if bits[1] {
+        scalar *= F::ZETA;
+    }
+
+    scalar
+}
+
+/// Maps a K-bit bitstring to a scalar.
+///
+/// This corresponds to Algorithm 1 from [BGH2019], where `F` corresponds to $F_q$, the
+/// scalar field of $P$. Where Algorithm 1 computes $Acc = [scalar] P$, this function
+/// computes `scalar`.
+///
+/// [BGH2019]: https://eprint.iacr.org/2019/1021.pdf
+#[allow(dead_code)]
+pub(crate) fn endoscale_scalar<F: FieldExt>(acc: Option<F>, bits: &[bool]) -> F {
+    assert_eq!(bits.len() % 2, 0);
+
+    let mut acc = if let Some(acc) = acc {
+        acc
+    } else {
+        (F::ZETA + F::one()).double()
+    };
+    for j in (0..(bits.len() / 2)).rev() {
+        let pair = [bits[2 * j], bits[2 * j + 1]];
+        let endo = endoscale_pair_scalar::<F>(pair);
+        acc = acc.double();
+        acc += endo;
+    }
+    acc
+}
+
+/// Maps a pair of bits to a multiple of a base using endoscaling.
+pub(crate) fn endoscale_pair<C: CurveAffine>(bits: [bool; 2], base: C) -> CtOption<C> {
```

Comment:
We don't currently have a non-identity type in either `group` or `pasta_curves`. We do have similar wrappers in the `orchard` crate, but we should see if there's a not-horrible way to upstream this into the `group` type system (the initial way I'd imagine integrating this would require downstream curve impls to provide wrapper types equivalent to what we implemented manually in `orchard`).